domain/application/service/unit.go

```diff
@@ -74,6 +108,34 @@ func (s *Service) AddUnits(ctx context.Context, appName string, units ...AddUnit
 	return nil
 }
 
+func (s *ProviderService) mergeApplicationAndModelConstraints(ctx context.Context, appUUID coreapplication.ID) (coreconstraints.Value, error) {
+	validator, err := s.constraintsValidator(ctx)
+	if err != nil {
+		return coreconstraints.Value{}, internalerrors.Capture(err)
+	}
+	// If the provider doesn't support constraints validation, then we can
+	// just return the zero value.
+	if validator == nil {
+		return coreconstraints.Value{}, nil
+	}
+
+	appCons, err := s.GetApplicationConstraints(ctx, appUUID)
+	if err != nil {
+		return coreconstraints.Value{}, internalerrors.Errorf("retrieving application %q constraints for merging with model constraints args: %w", appUUID.String(), err)
+	}
+	modelCons, err := s.st.GetModelConstraints(ctx)
+	if err != nil && !errors.Is(err, modelerrors.ConstraintsNotFound) {
+		return coreconstraints.Value{}, internalerrors.Errorf("retrieving model constraints for merging with application %q constraints: %w	", appUUID.String(), err)
+	}
```

Comment:
nit: How difficult would it be to drop the service call and call to state instead both times?

Reply:
Not that much, but I'll still need the validation that we have in the service layer, so I don't see the point of it.